# Advent of Code Day 3 - Gear Ratios

In [22]:
# Setup
from aocd.models import Puzzle

# To Clear Cache rm -rf ~/.config/aocd/github.yurifrl.XXXXX

# Get the input data
puzzle = Puzzle(day=3, year=2023)

# Get the imput and parse it
input = puzzle.input_data.split('\n')

#
[(example, (answerA, answerB))] = [(ex.input_data.split('\n'), (ex.answer_a, ex.answer_b)) for ex in puzzle.examples]
[(example[:1], (answerA, answerB))]

[(['467..114..'], ('4361', None))]

# First Part

You and the Elf eventually reach a gondola lift station; he says the gondola lift will take you up to the water source, but this is as far as he can bring you. You go inside.

It doesn't take long to find the gondolas, but there seems to be a problem: they're not moving.

"Aaah!"

You turn around to see a slightly-greasy Elf with a wrench and a look of surprise. "Sorry, I wasn't expecting anyone! The gondola lift isn't working right now; it'll still be a while before I can fix it." You offer to help.

The engineer explains that an engine part seems to be missing from the engine, but nobody can figure out which one. If you can add up all the part numbers in the engine schematic, it should be easy to work out which part is missing.

The engine schematic (your puzzle input) consists of a visual representation of the engine. There are lots of numbers and symbols you don't really understand, but apparently any number adjacent to a symbol, even diagonally, is a "part number" and should be included in your sum. (Periods (.) do not count as a symbol.)

Here is an example engine schematic:


```
467..114..
...*......
..35..633.
......#...
617*......
.....+.58.
..592.....
......755.
...$.*....
.664.598..
```

In this schematic, two numbers are not part numbers because they are not adjacent to a symbol: 114 (top right) and 58 (middle right). Every other number is adjacent to a symbol and so is a part number; their sum is 4361.

Of course, the actual engine schematic is much larger. What is the sum of all of the part numbers in the engine schematic?

## Summary

- Task: Calculate the sum of all "part numbers" in an engine schematic.
- Criteria: A "part number" is any number adjacent to a symbol (excluding periods) in the schematic.
- Example Result: In the provided example, sum of part numbers is 4361.
- Goal: Find the sum of part numbers in the actual, larger schematic.

In [23]:
def parser(input):
    external = []
    for i in input:
        internal = []
        for n in i:
            internal.append(n)
        external.append(internal)
    return external

In [21]:
def parse_schematic(grid):
    for r, row in enumerate(grid):
        for c, ch in enumerate(row):
            if ch.isdigit() or ch == ".":
                continue
            for cr in [r - 1, r, r + 1]:
                for cc in [c - 1, c, c + 1]:
                    print(cr, cc)
    return grid

parse_schematic(parser(example))

0 2
0 3
0 4
1 2
1 3
1 4
2 2
2 3
2 4
2 5
2 6
2 7
3 5
3 6
3 7
4 5
4 6
4 7
3 2
3 3
3 4
4 2
4 3
4 4
5 2
5 3
5 4
4 4
4 5
4 6
5 4
5 5
5 6
6 4
6 5
6 6
7 2
7 3
7 4
8 2
8 3
8 4
9 2
9 3
9 4
7 4
7 5
7 6
8 4
8 5
8 6
9 4
9 5
9 6


[['4', '6', '7', '.', '.', '1', '1', '4', '.', '.'],
 ['.', '.', '.', '*', '.', '.', '.', '.', '.', '.'],
 ['.', '.', '3', '5', '.', '.', '6', '3', '3', '.'],
 ['.', '.', '.', '.', '.', '.', '#', '.', '.', '.'],
 ['6', '1', '7', '*', '.', '.', '.', '.', '.', '.'],
 ['.', '.', '.', '.', '.', '+', '.', '5', '8', '.'],
 ['.', '.', '5', '9', '2', '.', '.', '.', '.', '.'],
 ['.', '.', '.', '.', '.', '.', '7', '5', '5', '.'],
 ['.', '.', '.', '$', '.', '*', '.', '.', '.', '.'],
 ['.', '6', '6', '4', '.', '5', '9', '8', '.', '.']]

# Second Part
TODO

## Summary
TODO

In [ ]:
def partB(input):
    return 0

answerB == partB(example)
answerB

In [ ]:
partB(input)

0